src/subscript/co2containment/co2_calculation.py

```diff
@@ -0,0 +1,820 @@
+                + 1 / (1000 * bgas[date])
+            ),
+        ]
+        co2_molar_vol[date][0] = [
+            0 if x < 0 or y == 0 else x
+            for x, y in zip(co2_molar_vol[date][0], xmf2[date])
+        ]
+        co2_molar_vol[date][1] = [
+            0 if x < 0 or y == 0 else x
+            for x, y in zip(co2_molar_vol[date][1], ymf2[date])
+        ]
+    return co2_molar_vol
+
+
+def _pflotran_co2_simple_volume(source_data: SourceData) -> Dict:
+    """
+    Calculates CO2 "simple" volume based on the existing properties in PFlotran
+
+    Args:
+      source_data (SourceData): Data with the information of the necessary properties
+                                for the calculation of CO2 "simple" volume
+
+    Returns:
+      Dict
+
+    """
+    dates = source_data.DATES
+    sgas = source_data.SGAS
+    ymfg = source_data.YMFG
+    amfg = source_data.AMFG
+    eff_vols = source_data.PORV
+    co2_vol_st1 = {}
+    for date in dates:
+        co2_vol_st1[date] = [
+            eff_vols[date] * (1 - sgas[date]) * amfg[date],
+            eff_vols[date] * sgas[date] * ymfg[date],
+        ]
+    return co2_vol_st1
+
+
+def _eclipse_co2_simple_volume(source_data: SourceData) -> Dict:
+    """
+    Calculates CO2 "simple" volume based on the existing properties in Eclipse
+
+    Args:
+        source_data (SourceData): Data with the information of the necessary properties
+                                  for the calculation of CO2 "simple" volume
+
+    Returns:
+        Dict
+
+    """
+    dates = source_data.DATES
+    sgas = source_data.SGAS
+    xmf2 = source_data.XMF2
+    ymf2 = source_data.YMF2
+    eff_vols = source_data.RPORV
+    co2_vol_st1 = {}
+    for date in dates:
+        co2_vol_st1[date] = [
+            eff_vols[date] * (1 - sgas[date]) * xmf2[date],
+            eff_vols[date] * sgas[date] * ymf2[date],
+        ]
+    return co2_vol_st1
+
+
+def _calculate_co2_data_from_source_data(
+    source_data: SourceData,
+    calc_type: CalculationType,
+    co2_molar_mass: float = DEFAULT_CO2_MOLAR_MASS,
+    water_molar_mass: float = DEFAULT_WATER_MOLAR_MASS,
+) -> Co2Data:
+    """
+    Calculates a given calc_type (mass/volume_extent/volume_actual/volume_actual_simple)
+    from properties in source_data.
+
+    Args:
+        source_data (SourceData): Data with the information of the necessary properties
+                                  for the calculation of calc_type
+        calc_type (CalculationType): Which amount is calculated (mass/volume_extent/
+                                     volume_actual/volume_actual_simple)
+        co2_molar_mass (float): CO2 molar mass - Default is 44 g/mol
+        water_molar_mass (float): Water molar mass - Default is 18 g/mol
+
+    Returns:
+      Co2Data
+    """
+    # pylint: disable-msg=too-many-locals
+    # pylint: disable-msg=too-many-branches
+    # pylint: disable-msg=too-many-statements

```

Comment:
Just a suggestion, you could consider refactor the code to have less branches so that it's easier to read.

Reply:
Probably a good idea, maybe do that sometime in the future.